### PR TITLE
Feature/add mosaic to fetch page

### PR DESCRIPTION
**Pull request by zdzislawbela**

### CODE REVIEW

**Comment by raqso** [comment]
image
You've got some regression :boom:

File: components/Buttons/MainButton/MainButton.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+import Link from "next/link";
+
+import ButtonBase from "@material-ui/core/ButtonBase";
+import Typography from "@material-ui/core/Typography";
+
+import { useStyles } from "./styles";
+
+type props = {
```
Comment (raqso) [inline components/Buttons/MainButton/MainButton.tsx]: Use PascalCase for types names

https://basarat.gitbook.io/typescript/styleguide#type

File: components/Buttons/MainButton/MainButton.tsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    <div className={classes.root}>
+      <Link href={href}>
+        <ButtonBase
+          focusRipple
+          key={title}
+          className={classes.image}
+          focusVisibleClassName={classes.focusVisible}
+          style={{
+            width: "80vw",
+          }}
```
Comment (raqso) [inline components/Buttons/MainButton/MainButton.tsx]: Extract to class. In such a way, the style object will be recreated on every render.

File: components/FetchedDogs/FetchedDogs.tsx [hunk 3]
```diff
@@ -1,119 +1,86 @@
...
     likedDogs,
     setLikedDogs,
     setIsError,
-    apiCallCounter,
-    setApiCallCounter,
+    storagedBreeds,
   } = useAppContext();
 
   const [breed, setBreed] = useState<Breed>();
   const [random, setRandom] = useState(0);
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: What does it mean?

File: components/FetchedDogs/FetchedDogs.tsx [hunk 4]
```diff
@@ -1,119 +1,86 @@
...
+  const [isScreenFilled, setIsScreenFilled] = useState(false);
   const observer = useRef<null | IntersectionObserver>(null);
 
-  const getRandomBreed = () => {
+  const getNextDog = () => {
     setRandom(Math.floor(Math.random() * storagedBreeds.length));
 
     if (random === 0 || random === 1) {
       setRandom(-1);
     }
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: I'd extract to function called `getBreed()` or something

Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: Why set this state 2 times? Maybe just put this line above and the current 32 line in the else?

File: components/FetchedDogs/FetchedDogs.tsx [hunk 5]
```diff
@@ -1,119 +1,86 @@
...
 
     setBreed(storagedBreeds[random]);
+
     return storagedBreeds[random];
   };
 
+  const waitForAll = () => {
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: Doesn't look stable

File: components/FetchedDogs/FetchedDogs.tsx [hunk 6]
```diff
@@ -1,119 +1,86 @@
...
       if (node) observer.current.observe(node);
     },
     [loading]
   );
 
-  const getPosition = (string: string, subString: string, index: number) => {
-    return string.split(subString, index).join(subString).length;
+  const fetchDog = () => {
+    setLoading(true);
+    useFetchDog(storagedBreeds[random]).then((newDog) => {
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: What is 'useFetchDog'? If it's a hook, it shouldn't be called here. If it's just a method, it shouldn't be named like a hook.

File: components/FetchedDogs/FetchedDogs.tsx [hunk 7]
```diff
@@ -1,119 +1,86 @@
...
+      setFetchedDogs([...fetchedDogs, newDog]);
+      setLoading(false);
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: Handle error, if an error will occur, the loading should be also changed to false.

File: hooks/useFetchDog.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
+import { DogDetails } from "../context";
+
+type fetchedDog = {
```
Comment (raqso) [inline hooks/useFetchDog.ts]: PascalCase

File: hooks/useFetchDog.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  const fetchedDog: fetchedDog = await response.json();
+
+  const newDogsDetails: DogDetails = {
+    status: fetchedDog.status,
+    breedName: fetchedDog.message.slice(
+      30,
+      getPosition(fetchedDog.message, "/", 5)
+    ),
+    image: fetchedDog.message,
+    time: timeInMs,
```
Comment (raqso) [inline hooks/useFetchDog.ts]: 'time' not much meaningful. You meant 'fetchedAt' maybe?

**Review by raqso — CHANGES_REQUESTED** [review]

File: components/FetchedDogs/FetchedDogs.tsx (hunk 3 above)
Comment (zdzislawbela) [reply components/FetchedDogs/FetchedDogs.tsx]: Not sure what's this question is pointing. Random state? If yes that was an old idea to keep random breed index tracking. Not necessary any more. I change it.

Comment (raqso) [reply components/FetchedDogs/FetchedDogs.tsx]: What name means, because `random` tells nothing. After what you wrote it should be more like `randomBreedIndex`

**Review by raqso — CHANGES_REQUESTED** [review]

File: components/FetchedDogs/FetchedDogs.tsx (hunk 5 above)
Comment (zdzislawbela) [reply components/FetchedDogs/FetchedDogs.tsx]: Needs to be replaced with proper condition...

File: hooks/useDogs.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+  const [storagedBreeds] = useLocalStorage<string[]>(
+    "selectedBreeds",
+    breedsData
+  );
+
+  const loadMore = async (howMany = 4, setEmpty = false) => {
+    if (setEmpty) {
+      return setDogs([]);
+    }
```
Comment (raqso) [inline hooks/useDogs.ts]: Doesn't feel right. I'd extract this as another method if needed, it has nothing to do with this one ('load more dogs').

File: hooks/useDogs.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    setLoading(true);
+
+    const getDog = async () => {
+      const randomIndex = getRandomArrayIndex(storagedBreeds);
+      const breedName = storagedBreeds[randomIndex];
+      const DOGS_API = "https://dog.ceo/api/breed/";
```
Comment (raqso) [inline hooks/useDogs.ts]: Such things should be kept globally in the project, not in nested functions scopes... Ideally make this configurable via env vars

File: hooks/useDogs.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+      const RANDOM_IMAGE = "/images/random";
```
Comment (raqso) [inline hooks/useDogs.ts]: Variable lies, there is no random image inside.

File: hooks/useDogs.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        downloadedAt: timeInMs,
+      } as DogDetails;
+    };
+
+    try {
+      const promises = Array(howMany)
+        .fill(null)
+        .map(() => {
+          return getDog();
+        });
```
Comment (raqso) [inline hooks/useDogs.ts]:
```suggestion
        .map(getDog);
```

File: hooks/useDogs.ts [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+      const newDogDetails = await Promise.allSettled(promises);
+
+      const fulfilledNewDogDetails = newDogDetails
+        .filter((promise) => promise.status === "fulfilled")
+        .map((fulfilledPromise) => {
+          //@ts-ignore
```
Comment (raqso) [inline hooks/useDogs.ts]: Maybe try something from here: https://marcinbiernat.pl/2020/02/Type-safe-filters/

File: hooks/useDogs.ts [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+          return fulfilledPromise.value as DogDetails;
+        });
+
+      setDogs((currentDogs) => [...currentDogs, ...fulfilledNewDogDetails]);
+    } catch (error) {
+      setError(error);
```
Comment (raqso) [inline hooks/useDogs.ts:71]: I'd also console.error this error here, to have e.g. call stack for debugging in the browser console.

File: hooks/useFetchDog.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import { DogDetails } from "../context";
```
Comment (raqso) [inline hooks/useFetchDog.ts]: Is this file needed at all?

File: pages/fetch.tsx [hunk 17]
```diff
@@ -6,11 +6,11 @@ import { useAppContext } from "../context";
...
 
 export default function Fetch() {
-  const { fetchedDogs } = useAppContext();
-  const numberOfDogs = fetchedDogs.length;
+  const { dogs } = useAppContext();
+  const numberOfDogs = dogs.length;
   let titleOfPage = `🐕 You fetched ${numberOfDogs} dogs!`;
 
-  if (fetchedDogs.length === 0) {
+  if (dogs.length === 0) {
```
Comment (raqso) [inline pages/fetch.tsx]: If you've got the loading state, I'd use that. You can have the case when loading has finished and there is 0 results and you will be displaying user that it's still loading (which is not true).

File: utils/getRandomArrayIndex.ts [hunk 18]
```diff
@@ -0,0 +1,3 @@
+export const getRandomArrayIndex = (array: string[]) => {
```
Comment (raqso) [inline utils/getRandomArrayIndex.ts]: It doesn't have to be string, it will be working for any elements.
```suggestion
export const getRandomArrayIndex = <T>(array: T[]) => {
```

File: data/breedsData.ts [hunk 19]
```diff
@@ -4,7 +4,7 @@ export const breedsData = [
   "airedale",
   "akita",
   "appenzeller",
-  "australian-shepherd",
+  "australian/shepherd",
```
Comment (raqso) [inline data/breedsData.ts]: Do not hardcode it like this. It's 100% dependent on API, so get list of possible breeds from API first.
1. So create an elegant hook like `useBreeds` and fetch/format everything inside.
2. You could also chek out awesome library to use instead of regular fetch which will take care of caching this for you: https://react-query.tanstack.com/

File: components/FetchedDogs/FetchedDogs.tsx [hunk 20]
```diff
@@ -1,121 +1,53 @@
...
-    setRandom(Math.floor(Math.random() * storagedBreeds.length));
-
-    if (random === 0 || random === 1) {
-      setRandom(-1);
-    }
-
-    setBreed(storagedBreeds[random]);
-    return storagedBreeds[random];
+  const waitForAll = () => {
+    const miliSeconds = isMosaic ? 6000 : 2500;
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: Magic numbers :mage:

**Review by raqso — CHANGES_REQUESTED** [review]

File: components/Buttons/MainButton/MainButton.tsx (hunk 2 above)
Comment (zdzislawbela) [reply components/Buttons/MainButton/MainButton.tsx]: image
it's example from material-ui itself. Or am I wrong?

https://v1.material-ui.com/demos/buttons/

File: components/FetchedDogs/FetchedDogs.tsx (hunk 20 above)
Comment (zdzislawbela) [reply components/FetchedDogs/FetchedDogs.tsx]: This workaround needs to be fully replaced.

File: components/Buttons/MainButton/MainButton.tsx (hunk 2 above)
Comment (raqso) [reply components/Buttons/MainButton/MainButton.tsx]: In the example, you've got dynamic value from an element, really. You're hardcoding text "80vh".

File: components/FetchedDogs/FetchedDogs.tsx [hunk 21]
```diff
@@ -24,7 +24,13 @@ export const FetchedDogs = () => {
   const observer = useRef<null | IntersectionObserver>(null);
 
   const waitForAll = () => {
-    const miliSeconds = isMosaic ? 6000 : 2500;
+    const defaultWaitingTimeInMls = 2500;
+    const mosaicWaitingTimeInMls = 6000;
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: It's a common pattern to name constants in ALL_CAPS_SNAKE_CASE

File: .env [hunk 22]
```diff
@@ -0,0 +1,2 @@
+NEXT_PUBLIC_DOGS_API_PREFIX = public_url_prefix
+NEXT_PUBLIC_DOGS_API_SUFFIX = public_url_suffix
```
Comment (raqso) [inline .env]: co

File: hooks/useDogs.ts [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+
+  const loadMore = async (howMany = 4) => {
+    setLoading(true);
+
+    const getDog = async () => {
+      const randomIndex = getRandomArrayIndex(storagedBreeds);
+      const breedName = storagedBreeds[randomIndex];
+
+      const response = await fetch(
+        `${process.env.NEXT_PUBLIC_DOGS_API_PREFIX}${breedName}${process.env.NEXT_PUBLIC_DOGS_API_SUFFIX}`
```
Comment (raqso) [inline hooks/useDogs.ts]: > process.env.NEXT_PUBLIC_DOGS_API_SUFFIX

I don't think it should be in envs

**Review by raqso — CHANGES_REQUESTED** [review]

File: hooks/useDogs.ts (hunk 23 above)
Comment (zdzislawbela) [reply hooks/useDogs.ts]: I guess you meant that?
image

Comment (raqso) [reply hooks/useDogs.ts]: Yes

File: .env.local.example [hunk 24]
```diff
@@ -0,0 +1,2 @@
+NEXT_PUBLIC_DOG_API = https://dog.ceo/api/breed/
+NEXT_PUBLIC_DOG_API_ALL_BREEDS = https://dog.ceo/api/breeds/list/all
```
Comment (raqso) [inline .env.local.example]: 😬

File: hooks/useBreeds.ts [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+  loadBreeds: () => Promise<void>;
+};
+
+export interface BreedsDetails {
+  message: { [key: string]: string[] };
+  status: string;
+}
+
+export const useBreeds = (): UseBreeds => {
+  const { setStoragedBreeds } = useAppContext();
```
Comment (raqso) [inline hooks/useBreeds.ts]: 😬

File: hooks/useBreeds.ts [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+          .flatMap((breed) => breed);
+      };
+
+      const breedsWithSubBreeds = concatenateBreedsSubBreeds();
+      return breedsWithSubBreeds;
+    };
+
+    try {
+      const promise = getBreeds();
+      const breeds = await Promise.resolve(promise);
```
Comment (raqso) [inline hooks/useBreeds.ts]: Why not just?
```suggestion
      const breeds = await getBreeds();
```

**Review by raqso — CHANGES_REQUESTED** [review]

File: .env.local.example (hunk 24 above)
Comment (zdzislawbela) [reply .env.local.example]: ?

Comment (raqso) [reply .env.local.example]: I don't think the second env is needed

File: hooks/useDogs.ts [hunk 27]
```diff
@@ -16,6 +16,11 @@ type UseDogs = {
   setEmpty: () => void;
 };
 
+interface PromiseFulfilledResult<T> {
```
Comment (raqso) [inline hooks/useDogs.ts:19]: :+1:

File: hooks/useBreeds.ts [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+    } catch (error) {
+      console.error(error);
+      setErrorBreeds(error);
+    } finally {
+      setLoadingBreeds(false);
+    }
+  };
+
+  useEffect(() => {
+    console.log("effect in useBreeds");
```
Comment (raqso) [inline hooks/useBreeds.ts]: :wastebasket:

File: components/FetchedDogs/FetchedDogs.tsx [hunk 29]
```diff
@@ -1,121 +1,59 @@
...
-  const [breed, setBreed] = useState<Breed>();
-  const [random, setRandom] = useState(0);
-  const [storagedBreeds] = useLocalStorage<Breed[]>("selectedBreeds", []);
-
+  const [isScreenFilled, setIsScreenFilled] = useState(false);
   const observer = useRef<null | IntersectionObserver>(null);
 
-  const getRandomBreed = () => {
-    setRandom(Math.floor(Math.random() * storagedBreeds.length));
+  const waitForAll = () => {
```
Comment (raqso) [inline components/FetchedDogs/FetchedDogs.tsx]: Is this even needed? Is see some issues here:
- the function is recreated every render
- **the function is called every render**
- **you setup a new setTimeout every render**
- **you don't clear already set timeouts**.
E.g. If the page will rerender 3 times, and I will go to another one, before timeout, those 3 setTimeouts will throw errors. It's due to the fact that the page component won't exist anymore so do the `setIsScreenFilled` method. Not to mention possible performance problems.

The only possible way would be to use this inside useEffect once and to clear already set timeout on component unmount. But as I said, probably you don't need this method at all, think about this.

**Review by raqso — CHANGES_REQUESTED** [review]

**Review by raqso — APPROVED** [review]
